index.bs

```diff
@@ -0,0 +1,550 @@
+  idlharness.https.window.js
+</wpt>
+
+### {{IdleDetector/userState}} attribute ### {#api-idledetector-userstate}
+
+The <dfn attribute for=IdleDetector>userState</dfn> attribute MUST be
+initialized to `null`.
+
+<div algorithm="user becomes idle">
+
+If |this|.{{[[state]]}} is `"started"`, |this|.{{userState}} is {{"active"}} and
+the user has not interacted with the device in at least |this|.{{[[threshold]]}}
+milliseconds, the [=user agent=] MUST [=queue a task=] to run the following
+steps:
+
+1.  Set |this|.{{userState}} to {{"idle"}}.
+1.  [=Fire an event=] named {{"change"}} using the {{Event}} interface at
+    |this|.
+
+<wpt>
+  interceptor.https.html
+</wpt>
+</div>
+
+<div algorithm="user becomes active">
+
+If |this|.{{[[state]]}} is `"started"`, |this|.{{userState}} is {{"idle"}} and
+the user interacts with the device in, the [=user agent=] MUST [=queue a task=]
+to run the following steps:
+
+1.  Set |this|.{{userState}} to {{"active"}}.
+1.  [=Fire an event=] named {{"change"}} using the {{Event}} interface at
+    |this|.
+
+<wpt>
+  interceptor.https.html
+</wpt>
+</div>
+
+### {{IdleDetector/screenState}} attribute ### {#api-idledetector-screenstate}
+
+The <dfn attribute for=IdleDetector>screenState</dfn> attribute MUST be
+initialized to `null`.
+
+<div algorithm="screen locked">
+
+If |this|.{{[[state]]}} is `"started"` and the device is locked by the user, an
+automatic timeout or any other event, the [=user agent=] MUST [=queue a task=] to
+run the following steps:
+
+1.  Set |this|.{{screenState}} to {{"locked"}}.
+1.  [=Fire an event=] named {{"change"}} using the {{Event}} interface at
+    |this|.
+
+<wpt>
+  interceptor.https.html
+</wpt>
+</div>
+
+<div algorithm="screen unlocked">
+
+If |this|.{{[[state]]}} is `"started"` and the device is unlocked by the user or
+for any other reason, the [=user agent=] MUST [=queue a task=] to run the
+following steps:
+
+1.  Set |this|.{{screenState}} to {{"unlocked"}}.
+1.  [=Fire an event=] named {{"change"}} using the {{Event}} interface at
+    |this|.
+
+<wpt>
+  interceptor.https.html
+</wpt>
+</div>
+
+### {{IdleDetector/onchange}} attribute ### {#api-idledetector-onchange}
+
+<dfn attribute for="IdleDetector">onchange</dfn> is an <a>Event handler IDL
+attribute</a> for the {{change}} event type.
+
+### {{IdleDetector/requestPermission()}} method ### {#api-idledetector-requestpermission}
+
+<div algorithm>
+The <dfn method for=IdleDetector>requestPermission()</dfn> method, when invoked, MUST run these steps:
+
+1.  Let |result| be [=a new promise=].
+1.  Run these steps [=in parallel=].
+    1.  If the [=relevant global object=] of [=this=] does not have [=transient
+        activation=], [=reject=] |result| with {{NotAllowedError}} and abort
+        there steps.
+    1.  Let |permissionState| be the result of [=requesting permission to use=]
+        the [=powerful feature=] named `"idle-detection"`.
+    1.  Resolve |result| with |permissionState|.
```

Comment:
You need to queue a task to resolve the promise these days. (For some time in specs it would queue a task for you automatically, but we realized that was a bad idea, as it did not properly specify the task source/event loop/document.)

Reply:
Hmm, this is a bit tricky.

I think the activation check should probably be synchronous, not inside the task.

Then, I think the permission checking needs to get a bit more complicated, since as written it is posting a task which, when excecuted, will synchronously block the entire event loop while asking the user for permission.

So I think the structure desired is:

1. If no user activation, return a promise rejected with...
1. Let result be a new promise.
1. In parallel:
   1. Let permissionState be the result of...
   1. Queue a global task on the relevant global object of this using the XXX task source to resolve result with permissionState.
1. Return result.